From 670747fedbcd78f45d45387b3ec1833b2c328be2 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zahari.doychev@intel.com>
Date: Fri, 19 Jun 2020 11:13:08 +0200
Subject: [PATCH] pon_qos: fix CPPU queues lookup table mappings

The CPU port queue mappings should be based on the tc classid e.g. classid is
8000:2 then the queue belonging to 8000:2 has to be mapped traffic class 9.
The general rule is classid X:Y -> TC = (Y-1) + 8. This is because OMCI
packets are marked with TC=8 and go to the high priority queue.

Signed-off-by: Zahari Doychev <zahari.doychev@intel.com>
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index 4f5dd6afd44e..c4cfc65b0569 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -911,6 +911,14 @@ int pon_qos_update_qmap(struct net_device *dev,
 
 	/* TODO: port zero is currently the CPU port */
 	if (!sch->port) {
+		/* CPU TC mappings are based on the class id handle for now i.e.
+		 * queue with classid X:Y is mapped tc = (Y-1) + 8.
+		 * Later this can be improved by using tc-cookie.
+		 */
+		tc = TC_H_MIN(q_tc->handle) - 1;
+		if (tc < 0 || tc > PON_QOS_8021P_HIGHEST_PRIO)
+			return -EINVAL;
+
 		ret = pon_qos_config_cpu_port(qid, sch->port, tc, en);
 	} else {
 		if (sch->ds && q_tc->indev) {
